### PR TITLE
Basic mixed blocks for `float#` in runtime 5

### PR DESCRIPTION
**Pull request by ncik-roberts**
Support records with `float#` fields in addition to value fields, and represent the unboxed fields as flat at runtime. This PR only adds support for the OCaml 5 runtime, not the OCaml 4 runtime.

Terminology:
  * The record declaration with mixed value and unboxed fields is called a "mixed record".
  * The runtime representation of a record with `float#` and value fields is called a "mixed block".

The native code compiler and bytecode compiler support mixed records. Native code represents mixed record values as mixed blocks; bytecode represents mixed record values as normal blocks.

A mixed record must consist of a **value prefix** (a sequence of fields whose layout is value) followed by a **flat suffix** (a sequence of fields that are either immediates or unboxed floats). This restriction mirrors the mixed block runtime representation, where the 8 top bits of the header are reserved to encode the length of the flat suffix. The length of the flat suffix is read by the GC and other polymorphic traversals, both to detect whether the block is a mixed block and to decide how to recursively descend into the fields.

We use the "reserved header bits" configuration setting to steal 8 bits for the tag.

## What is supported in this PR

* Mixed records where the only non-value fields are `float#`, compiled as mixed blocks in native code.
* Records whose fields are all `float` or `float#`. These are encoded as mixed blocks with scannable prefix 0, and all fields are stored flat.

## What is not supported in this PR (that we plan to support soon)

* Cleanup for reserving header bits.
* Support in the OCaml 4 runtime.
* Inline record arguments to constructors.
* `int64#`, `nativeint#`, and `int32#` fields

## What is not supported in this PR (that we don't have immediate plans to build support for)

* Polymorphic comparison, hash, and marshaling

## How to review

Review directory-by-directory, not commit-by-commit.

  * @ccasin, please review `ocaml/typing/`, `ocaml/utils/`, and `ocaml/lambda/`. I would also appreciate your input on all of the tests, not just the ones relevant to type-checking.
  * @TheNumbat, please review `middle_end/flambda2` and `backend/cmm_helpers.ml`.
  * I will figure out a good person for the runtime and bytecode changes.

Many tests are auto-generated. I check in a small example of the auto-generated test for review.

### CODE REVIEW

**Comment by ncik-roberts** [comment]
The tests are failing because you need to pass `--enable-reserved-header-bits=8` to enable mixed block support. The full invocation is:

```
./configure --enable-reserved-header-bits=8 --enable-runtime5 --disable-naked-pointers
```

**Comment by stedolan** [comment]
> Records whose fields are all float or float#. These are encoded as mixed blocks with scannable prefix 0, and all fields are stored flat.

> What is not supported in this PR: Polymorphic comparison, hash, and marshaling

Not supporting polymorphic compare for code using the new unboxed types seems fine. However, not supporting polymorphic compare for records containing `float` (rather than `float#`) is a regression. Does this PR break `Complex.one = Complex.one`?

> The tests are failing because you need to pass --enable-reserved-header-bits=8 to enable mixed block support

If the compiler doesn't work any more without reserved header bits, should we remove this as a configure option?

**Comment by ncik-roberts** [comment]
> Not supporting polymorphic compare for code using the new unboxed types seems fine. However, not supporting polymorphic compare for records containing float (rather than float#) is a regression.

Sorry, my summary was misleading. All-`float` records continue to be treated exactly the same.

This comment from my original summary is wrong:

> Records whose fields are all float or float#. These are encoded as mixed blocks with scannable prefix 0, and all fields are stored flat.

It should instead be about "records whose fields are all float or float#, and at least one field is float#".

**Comment by ncik-roberts** [comment]
> If the compiler doesn't work any more without reserved header bits, should we remove this as a configure option?

That seems like the best option to me. I'm working on this in another PR, and we should merge that at the same time as this.

File: ocaml/runtime/obj.c
Comment (stedolan) [inline ocaml/runtime/obj.c:176]: I think those `Bp_val` should be `Op_val`, since `scannable_sz` is in words, not bytes.

File: ocaml/runtime/minor_gc.c
Comment (stedolan) [inline ocaml/runtime/minor_gc.c:172]: Leftover debug code.

**Comment by stedolan** [comment]
When playing with this, I got a segfault when trying to print a value using the native toplevel (`ocamlnat`):
```
# type mx = { a : string; b : float# } ;;
type mx = { a : string; b : float#; }
# let x = { a = "a"; b = #1.0 };;
Segmentation fault
```
I think you need to update the `Type_record` case of `tree_of_val` in `ocaml/toplevel/genprintval.ml`.

File: ocaml/runtime/alloc.c
Comment (stedolan) [inline ocaml/runtime/alloc.c:350]: `CAML_STATIC_ASSERT`, even?

File: ocaml/runtime/minor_gc.c
Comment (stedolan) [inline ocaml/runtime/minor_gc.c:320]: The logic here took me a while to figure out. I don't think it's wrong, but I think it might be simpler to copy the non-scannable fields directly in oldify_one, as is currently done for entirely non-scannable blocks (where `tag >= No_scan-tag`).

That is, I think it would be simpler to directly copy the fields in `[scannable_sz, sz)`, and do the `sz > 1` tail-call test on `scannable_sz` instead (handling the `0` case with a return as you do here). That way, I think you can delete the new loop in `oldify_mopup`.

**Review by stedolan — CHANGES_REQUESTED** [review]
Looks good! I only read the runtime bits.

File: ocaml/typing/typedecl.ml
Comment (Ekdohibs) [inline ocaml/typing/typedecl.ml:1221]: Why do we use a mixed record there instead of a flat float record with `tag = 252`?

File: middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml
Comment (TheNumbat) [inline middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml:1435]:
```suggestion
    check_non_negative_imm imm "Psetmixedfield";
```

File: middle_end/flambda2/simplify/simplify_variadic_primitive.ml
Comment (TheNumbat) [inline middle_end/flambda2/simplify/simplify_variadic_primitive.ml:117]: Going by the following code, looks like you can construct the same fields list and use the match on mutable_or_immutable for both cases?

File: middle_end/flambda2/terms/flambda_primitive.mli
Comment (TheNumbat) [inline middle_end/flambda2/terms/flambda_primitive.mli:118]: Either add the tag/shape here or note that the fields are not necessary because they're only used for printing

File: middle_end/flambda2/to_cmm/to_cmm_primitive.ml
Comment (TheNumbat) [inline middle_end/flambda2/to_cmm/to_cmm_primitive.ml:104]: This and the below CR should be fine, at least until we allow packed int32#s, etc.

Note `unboxed_float_array_ref` always generates a mutable load, which is overly conservative. (It doesn't matter much)

**Review by TheNumbat — COMMENTED** [review]
I read `backend/` and `middle_end/flambda2`; looks good.
I'm not sure about the CRs in `middle_end/flambda2/simplify`, so an flambda2 author should also take a look.

File: middle_end/flambda2/simplify/simplify_variadic_primitive.ml
Comment (lthls) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml:117]: Yes, `T.any_block` is meant for blocks with regular constructors. Until we have support for mixed blocks in the Flambda2 type system, you should use `T.any_value`.

File: ocaml/typing/typedecl.ml
Comment (ncik-roberts) [reply ocaml/typing/typedecl.ml:1221]: First, why even support this case at all? The thinking is that this gives users a way of gradually migrating their all-float records to all-unboxed-floats without sacrificing the flat float optimization at any point.

Next, what tag to use?
  * Downside of current approach: the user loses polymorphic operations. But this is fine for Jane Street code.
  * Benefit of current approach: this could allow us to store boxed floats as flat even in inline record arguments to constructors. But actually I don't plan on supporting this: The argument for flat floats in mixed records doesn’t apply, as no existing code can possibly rely on the flat float optimization in constructor arguments.

So it does seem a bit better to use `Double_array_tag` here. The main reason I didn't do that in this PR is that @ccasin tells me it was hard to teach the middle-end how to do projections correctly here (some require boxing and others don't), and would require a small refactor.

But I plan on looking into this (in another PR) and suggest going with this approach for now.

**Comment by ncik-roberts** [comment]
> When playing with this, I got a segfault when trying to print a value using the native toplevel (`ocamlnat`):
> 
> ```
> # type mx = { a : string; b : float# } ;;
> type mx = { a : string; b : float#; }
> # let x = { a = "a"; b = #1.0 };;
> Segmentation fault
> ```
> 
> I think you need to update the `Type_record` case of `tree_of_val` in `ocaml/toplevel/genprintval.ml`.

@stedolan, fixing this (now-fixed) bug revealed to me another bug: `caml_hash` now can raise, and so external declarations thereof shouldn't be annotated as `[@@noalloc]`. I'm following up internally about alternative approaches for hashing.

File: ocaml/runtime/obj.c
Comment (ncik-roberts) [reply ocaml/runtime/obj.c:176]: Indeed.

File: ocaml/runtime/compare.c
Comment (ncik-roberts) [reply ocaml/runtime/compare.c:214]: This mirrors what we do for other non-comparable things, like abstract blocks and functions. It's admittedly strange but it's consistent.

**Review by ccasin — COMMENTED** [review]
I've read typing, lambda, and the tests for typing (but not the runtime tests).

This looks good!  Mainly minor comments/suggestions.

File: ocaml/typing/typedecl.ml
Comment (TheNumbat) [reply ocaml/typing/typedecl.ml:1]: no u

File: ocaml/runtime/minor_gc.c
Comment (ncik-roberts) [reply ocaml/runtime/minor_gc.c]: Done! I like this structure more, too.

File: backend/cmm_helpers.ml
Comment (ncik-roberts) [reply backend/cmm_helpers.ml:1490]: Yes, immediates are counted in the value section.

I actually don't think that we should call this scannable/non-scannable. The difference we're getting at here is the difference in sizes between values/floats, not whether or not the GC scans them.

File: middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml
Comment (ncik-roberts) [reply middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml:1435]: Accepted (manually).

File: middle_end/flambda2/simplify/simplify_variadic_primitive.ml
Comment (ncik-roberts) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml]: @TheNumbat: I'm not sure I understand your suggestion. I'm taking Vincent's suggestion to use `T.any_value` here; let me know if it seems like I'm missing something.

Comment (lthls) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml]: By the way, do you have a dedicated primitive for projecting regular values out of mixed blocks or do you use the existing `Pfield` primitive ? If you have a dedicated primitive then we can postpone any changes to the Flambda2 type system, if you reuse the existing primitive we will likely have to do at least some quick fixes to avoid wrongly confusing a mixed block and a regular block.

File: ocaml/testsuite/tests/translprim/comparison_table.heap.reference
Comment (ncik-roberts) [reply ocaml/testsuite/tests/translprim/comparison_table.heap.reference]: Marking resolved as the change I made (and have now reverted) is irrelevant to this PR.

File: ocaml/typing/typeopt.ml
Comment (ncik-roberts) [reply ocaml/typing/typeopt.ml:599]: The comment `CR mixed blocks v1` indicates that we'll need to make a change before v1. V1 will include non-tag-0 mixed blocks.

File: ocaml/typing/typedecl.ml
Comment (ncik-roberts) [reply ocaml/typing/typedecl.ml:3319]: I added a test for this now.

File: ocaml/typing/typedecl.ml
Comment (ncik-roberts) [reply ocaml/typing/typedecl.ml:1251]: I did the naive refactor (pull out the function and turn its free variables into parameters) and I don't think it's more readable. I did a smaller refactor (just pull out `find_flat_suffix`) and think that's much more readable. So that's what I did. Thanks for the suggestion.

File: middle_end/flambda2/simplify/simplify_variadic_primitive.ml
Comment (ncik-roberts) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml]: Per discussion, I'm going to change this PR to use a dedicated primitive for projecting a regular value out of a mixed block.

Comment (ncik-roberts) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml]: I now use the mixed-block version of the primitive for getting/setting fields in the value prefix as well.

File: ocaml/testsuite/tests/typing-layouts-float64/unboxed_floats_alpha.ml
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-float64/unboxed_floats_alpha.ml]: I split out two sets of tests:
  * Some tests under `ocaml/testsuite/tests/typing-layouts-float64` of all-float-and-float# mixed records
  * Some tests under `ocaml/testsuite/tests/mixed-blocks/test_mixed_blocks.ml` that test normal mixed records. Along your suggestion, I included a `string` field to more plainly suggest that the prior `float` fields are flat.

File: ocaml/testsuite/tests/typing-layouts-float64/basics.ml
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-float64/basics.ml]: Thanks, this was a dangling comment from an earlier version of this PR. It's now corrected.

File: ocaml/testsuite/tests/typing-layouts-float64/basics_alpha.ml
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-float64/basics_alpha.ml:315]: I've left this test in `typing-layouts-float64`, and added the other suggested tests in `typing-layouts/mixed_records.ml` (which I moved from `mixed-blocks/mixed_records.ml` and added a few more tests to).

File: ocaml/lambda/translcore.ml
Comment (ncik-roberts) [reply ocaml/lambda/translcore.ml]: Yeah, this was buggy and accidentally omitted from the tests. `typecore.ml` now supports it (and the auto-generated tests now exercise this too).

File: ocaml/toplevel/genprintval.ml
Comment (ncik-roberts) [reply ocaml/toplevel/genprintval.ml:545]: Indeed, that was segfaulting in a bunch of tests and is now fixed.

File: ocaml/lambda/lambda.mli
Comment (ncik-roberts) [reply ocaml/lambda/lambda.mli:778]: Added a comment.

File: ocaml/typing/types.mli
Comment (ncik-roberts) [reply ocaml/typing/types.mli:846]: accepted

File: ocaml/typing/typedecl.ml
Comment (ncik-roberts) [reply ocaml/typing/typedecl.ml]: Indeed. Consider it `ld`.

File: ocaml/typing/typedecl.ml
Comment (ncik-roberts) [reply ocaml/typing/typedecl.ml:1151]: Indeed, `flat` is redundant. I've renamed this to `Imm_element`.

File: ocaml/testsuite/tests/typing-layouts-float64/basics_alpha.ml
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-float64/basics_alpha.ml]: I ended up moving more of these tests into `mixed_records.ml`, which now has a twin `mixed_records_alpha.ml`. I minted new files instead of reusing `basics.ml` because the output for some tests is different in runtime 4. (To sidestep this, `mixed_records*.ml` is currently just run for runtime 5.)

File: ocaml/runtime/minor_gc.c
Comment (stedolan) [inline ocaml/runtime/minor_gc.c:315]: For multicore performance, this loop should be below the `if (try_update_object_header ...)`.

With multiple domains, if a block is reachable from multiple domains during minor GC then both will race to promote it, and `try_update_object_header` will return true only on the domain that wins the race. So, moving the loop down helps in two ways:

  - you are more likely to win the race if you don't stop to do a memcpy before crossing the finish line
  - if you lose the race, you don't need to do the copy at all

File: ocaml/runtime/minor_gc.c
Comment (stedolan) [inline ocaml/runtime/minor_gc.c:426]: `v` comes from `todo_list`, and blocks are now added to `todo_list` only if `scannable_wosize > 1`, so I think this condition is redundant.

**Review by stedolan — CHANGES_REQUESTED** [review]
Runtime stuff looks good! Two more small comments for `minor_gc.c`, but the rest lgtm.

File: ocaml/runtime/minor_gc.c
Comment (ncik-roberts) [reply ocaml/runtime/minor_gc.c]: I accepted this suggestion.

File: middle_end/flambda2/simplify/simplify_variadic_primitive.ml
Comment (ncik-roberts) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml:111]: I misunderstood @lthls 's earlier comment. I think this should be `any_value`, so I'll change that.

Comment (lthls) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml:111]: Yes, I confirm that `any_value` is the expected type for mixed block for the moment.

Comment (ncik-roberts) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml]: None of the tests were failing with the old code. Any suggestions for a test that could show the issue?

File: ocaml/runtime/minor_gc.c
Comment (ncik-roberts) [reply ocaml/runtime/minor_gc.c]: Indeed. I changed this to a `CAMLassert`. The interplay between `oldify_one` and `oldify_mopup` has confused me once again so I tried to write an explanatory comment on the assert.

File: ocaml/runtime/minor_gc.c
Comment (ncik-roberts) [reply ocaml/runtime/minor_gc.c]: (My naive attempt to move the `memcpy` down introduced a bug, because I forgot that `try_update_object_header` clobbers the 0th field of `v`. I've just pushed a fix.)

File: ocaml/typing/rec_check.ml
Comment (ccasin) [reply ocaml/typing/rec_check.ml]: For posterity: Nick ended up pushing a more fine-grained version of this that treats the flat suffix as `Dereference` and the value prefix as `Guard`.  I've reviewed this.

**Review by ccasin — APPROVED** [review]
Looks good now, thanks!

File: middle_end/flambda2/simplify/simplify_variadic_primitive.ml
Comment (ncik-roberts) [reply middle_end/flambda2/simplify/simplify_variadic_primitive.ml]: (resolving this for now as I'm not blocked on such a test.)

**Comment by ncik-roberts** [comment]
@stedolan I believe this is waiting for your approval as you requested changes earlier. Also, the latest commit includes one runtime change: the implementation of polymorphic hash for mixed blocks. Mind looking at that (and also at whether the other conversations are resolved to your liking)?

File: ocaml/runtime/hash.c
Comment (xclerc) [inline ocaml/runtime/hash.c:292]: (I don't think float values should be mixed
like that, `caml_hash_mix_double` is
performing some kind of normalization.)

Comment (ncik-roberts) [reply ocaml/runtime/hash.c:292]: In this PR, we could call `caml_hash_mix_double`. But in future PRs, there will be flat elements that are, say, `int64#` (or two `float32#`s packed together, or 4 `float16#s`, or a `float32#` followed by 2 `float16#s`). I'm worried there's not going to be a way to normalize these correctly.

* If we just normalize -0 to 0, then I think we could do that fairly easily even with arbitrary mixing of small floats.
* But I'm worried about NaN normalization. (Is that even needed? Two nans are never equal, so it's vacuously true that `nan1 = nan2` implies `hash(nan1) = hash(nan2)`.

Comment (xclerc) [reply ocaml/runtime/hash.c:292]: Sure, I think it is just safe to ignore that part of the block.

Comment (ncik-roberts) [reply ocaml/runtime/hash.c]: After some discussion:
  * We actually don't need to do any normalization. That's because mixed blocks can never "compare as equal" &mdash; polymorphic compare raises &mdash; so there's no risk of defying the hash invariant.
  * But this is all very tricky. I'm just going to return a constant for now on mixed blocks and revisit this later.

Comment (ncik-roberts) [reply ocaml/runtime/hash.c]: I'd like to change this to raise before moving mixed blocks out of `layouts_alpha.` But I'd like to do this in a separate PR to more easily assess performance impact of making `caml_hash` to not be `[@@noalloc]`. @stedolan would you mind reviewing the remaining runtime changes and approving if you think they look good (modulo the hash matter)?

**Review by stedolan — APPROVED** [review]

File: ocaml/runtime/minor_gc.c
Comment (stedolan) [reply ocaml/runtime/minor_gc.c]: Well spotted on the bug, I hadn't thought of that!

**Pull request merged**
